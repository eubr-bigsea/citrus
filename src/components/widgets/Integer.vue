<template>
    <div>
        <div v-if="readOnly">
            {{value === null ? field.default: value}}
        </div>
        <div v-else>
            <LabelComponent :field="field" :value="value" :show-help="showHelp"></LabelComponent>
            <input type="number" maxlength="10" class="form-control input-sm"
                :value="value === null ? field['default']: value" pattern="\\d*" @input="updated"
                onkeypress="return this.value.length <=15" :class="{'form-control-sm': small, 'w-25': small}"
                :required="field.required" />
        </div>
    </div>
</template>
<script>
    import LabelComponent from './Label.vue';
    import Widget from '../../mixins/Widget.js';
    export default {
        mixins: [Widget],
        components: { LabelComponent },
        methods: {
            updated: _.debounce(function (e) {
                //this.$root.$emit(this.message, this.field, e.target.value);
                this.triggerUpdateEvent(this.message, this.field, e.target.value);
            } , 500)
        },
        mounted() {
            const value = (this.field['default'] ? parseInt(this.field['default']) : null)
            this.$root.$emit(this.message,
                this.field, this.value || value);
        },
    }
</script>