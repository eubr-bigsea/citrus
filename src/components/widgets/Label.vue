<template>
    <span class="label">
        <span :data-order="field.order" :data-name="field.name"
            :class="{'text-danger': error, required: field.required}" :title="error">{{field.label || field.name}}</span>
        <span v-show="field.required">*</span>:
        <span v-if="(showHelp === undefined || showHelp) && field.help" class="data-help">
            <span class="fa fa-question-circle float-right" :title="field.help"></span>
        </span>
    </span>
</template>

<script>
    import Widget from '../../mixins/Widget.js';
    export default {
        mixins: [Widget],
        name: "property-label",
        props: {
            field: { required: true, type: Object },
            value: {},
            showHelp: true,
        },
        computed: {
            error() {
                if (this.field.required && (this.value === undefined || this.value === null || this.value === '' || this.value === {})) {
                    return this.$tc('errors.missingRequiredValue');
                }
                return null;
            },
        }
    }
</script>
<style scoped>
    .label {
        font-size: 1em;
    }
</style>
