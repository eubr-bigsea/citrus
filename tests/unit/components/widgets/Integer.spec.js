import { shallowMount } from '@vue/test-utils';
import Integer from '$SRC/components/widgets/Integer.vue';
import Label from '$SRC/components/widgets/Label.vue';
import { expect } from 'vitest';

describe('Integer.vue', () => {
    afterEach(() => {
        vi.restoreAllMocks();
        vi.useRealTimers();
    });
    beforeEach(() => {
        vi.useFakeTimers({ toFake: ['setTimeout', 'clearTimeout', 'Date'] });
    });
    it('renders read-only widget with default value', () => {
        const wrapper = shallowMount(Integer, {
            propsData: {
                field: {
                    value: null,
                    default: 1984
                },
                readOnly: true,
                value: null,
            }
        });
        expect(wrapper.findComponent(Label).exists()).toBe(false);
        expect(wrapper.text()).toBe("1984");

        //expect(wrapper.text()).toBe('Teste');
        //expect(wrapper.find('[class="label"]').exists()).toBeTruthy();
        //expect(wrapper.find("span[class='label']").exists()).toBeTruthy();
        //expect(wrapper.find("span[class='label']").exists()).toBeTruthy();

    });
    it('renders with field value', () => {
        const wrapper = shallowMount(Integer, {
            propsData: {
                field: {
                    value: 3001,
                    default: 0
                },
                readOnly: false,
                value: null
            }
        });
        expect(wrapper.findComponent(Label).exists()).toBeTruthy();
        expect(wrapper.text()).toBe("");
        const inputNumber = wrapper.find('input[type="number"]');

        expect(inputNumber.exists()).toBeTruthy();
        expect(inputNumber.element.pattern).toBe('\\\\d*');
        // console.debug(wrapper.html());
    });
    it('renders and emits after input ', async () => {
        const spy = vi.spyOn(Integer.methods, 'updated');
        const wrapper = shallowMount(Integer, {
            propsData: {
                field: {
                    value: 3001, default: 0
                },
                readOnly: false, value: null
            }
        });

        const inputNumber = wrapper.find('input');
        inputNumber.setValue(5000);
        // inputNumber.trigger('input');
        vi.advanceTimersByTime(1000);
        //await wrapper.vm.$nextTick();

        const payload = wrapper.emitted();
        // const spy = vi.spyOn(wrapper.vm, 'updated');
        // spy.mockImplementationOnce((e) => 'access-restricted');
        // expect(wrapper.vm.updated()).toEqual('access-restricted');
        // expect(spy.getMockName()).toEqual('updated');
        // console.debug(wrapper.vm.updated);
        // console.debug(wrapper.find('input[type="number"]').element.value);
        // console.table(payload)
        // console.table(payload.update[0][0])

        expect(payload.update[0][1]).toBe('5000');
        expect(spy).toBeCalled();

    });
    it('renders but with invalid value (non-integer) ', async () => {
        const spy = vi.spyOn(Integer.methods, 'updated');
        const wrapper = shallowMount(Integer, {
            propsData: {
                field: {
                    value: 'invalid number', default: 0
                },
                readOnly: false, value: null
            }
        });

        const inputNumber = wrapper.find('input');
        inputNumber.setValue('invalid');
        vi.advanceTimersByTime(1000);
        expect(inputNumber.element.value).toBe('');

        const payload = wrapper.emitted();
        expect(payload.update[0][1]).toBe('');
        expect(spy).toBeCalled();

    });
    it('snapshots the component', () => {
        const wrapper = shallowMount(Integer, { propsData: { field: {} } });
        expect(wrapper.text()).toMatchSnapshot();
    });
});
