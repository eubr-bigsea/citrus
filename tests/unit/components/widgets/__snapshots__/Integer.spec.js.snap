// Vitest Snapshot v1

exports[`Integer.vue > snapshots the component 1`] = `""`;
